### ISSUE TITLE
Add support for installing wheels

### ISSUE THREAD

**Issue by pombredanne** [buildout/buildout#144]
What would it take? a new zc.recipe.wheel I guess?

**Comment by mauritsvanrees** [comment]
Would indeed be nice if buildout supported wheels, mostly because presumably they are a lot faster to install.
I do not really have a good idea what would need to happen for this. I guess the easy_install part of zc.buildout would need to be updated to support wheels.
Or replace this easy_install part by pip, either a shipped version or an actual dependency. Does that make sense?
(Sorry, I am not volunteering. It looks outside of my skill set.)

**Comment by pombreda** [comment]
@mauritsvanrees I no longer use buildout because wheels are an order of magnitude faster to install. There is a built-in buildout egg dir to wheel converted that I contributed to wheel, so the conversion is rather easy based on some as-installed eggs. wheel or pip alone build wheels too.

**Comment by jimfulton** [comment]
@pombredanne It's interesting that wheels would be an order of magnitude faster to install, given that both wheels and eggs are just zip files. Do you know if this is really an egg vs wheel issue, or a difference between pip and buildout?

**Comment by aclark4life** [comment]
Not sure about speed, but Buildout should support:

- Finding wheels on package index
- Installing them according to https://www.python.org/dev/peps/pep-0427/#installing-a-wheel-distribution-1-0-py32-none-any-whl

Maybe the latter is challenging for Buildout as its approach is to provide scripts with sys.path loaded.

**Comment by pombredanne** [comment]
@jimfulton I think the reason is whether you are executing a setup.py or not. 
Pre-built wheels are just unzipped once. Installation is essentially just unzipping and no code is executed. 
In contrast there are few pre-built eggs available on Pypi and so most eggs are installed from a source distribution incurring the setup.py costs over and over.
So yes, the speedup is not because of the format per-se, but rather because of the pre-building.
Setuptools may also do many more checks on eggs than pip does of wheels, and that could be because of the format.

Practically installing 100 prebuilt wheels may take seconds whereas installing 100 source dists will take many many minutes. NB: I did not compare install speed between pre-built eggs and wheels... and I do not intend to do so.

**Comment by leorochael** [comment]
AFAICS, In Theory (TM), distlib provides everything we need to support both finding and installing wheels, including scripts:

http://pythonhosted.org/distlib/tutorial.html#installing-from-wheels

The wheel installation APIs are sufficiently flexible to allow installing anywhere, including, for example, an independent per-wheel directory, which is the way buildout likes to roll.

The `ScriptMaker` API also seems sufficiently overridable for `zc.recipe.egg` to subclass it to add all the per-wheel-directory sys.path extension code plus other `zc.recipe.egg` options.

The `wheel` package provides `bdist_wheel` command for `setuptools` to build wheels out of sdists.

The way I see buildout supporting wheels is by going all the way:

 * `distlib` is used to find and download sdists and wheels
 * `setuptools` with `wheel` is used to `bdist_wheel` the sdists
 * install the wheels found or created above into per-distribution directories inside a `wheel-directory` (successor to `egg-directory`)
 * `zc.recipe.egg` will write scripts the usual way, except putting wheel paths in `sys.path`, instead of egg paths.

The above means dropping support for installing from pre-built eggs found on pypi (or private indexes). if we want to support them, then:

 * A `distlib.locator.Locator` subclass adds `egg` to `.downloadable_extensions` [1]
 * `wheel convert` (or the `wheel.egg2wheel` API) should be used to convert eggs to wheels.

The wheel conversion support could be used also as part of a migration support from an egg based buildout to a wheel based one. This support could perhaps be packaged as a buildout extension so we don't pollute the buildout code with it.

This would obviously entail creating a buildout 3 release.

**Comment by aclark4life** [comment]
@leorochael I envision a ``wheels-directory`` too, but not to replace ``eggs-directory``. Buildout 3 should support both for the foreseeable future.

**Comment by WouterVH** [comment]
should this ticket be reopened?
I would love to see buildout supporting wheels, 
since this could mean we can use buildout much more in situations where gcc is not available.

**Comment by gforcada** [comment]
Just as a matter of reference, this last week I was trying to install a distribution (flake8-string-format) within plone.recipe.codeanalysis but it didn't work because it was only released as a wheel, I had to ask for a sdist release (fortunately the maintainer was quick on that).

So I guess that it will be more and more frequent that pure python distributions just release as wheels.

**Comment by sheilatron** [comment]
+1

**Comment by lfyzjck** [comment]
+1

I just found some library just provide wheel packages and buildout will not work if you use them.

For example: https://pypi.python.org/simple/httpsig/

**Comment by youngking** [comment]
+1

I want to install google's TensorFlow, but they only provide wheel packages.

**Comment by leorochael** [comment]
I've added some information to the [respective `setuptools` issue](https://github.com/pypa/setuptools/issues/78) that should allow anyone with enough round 'tuits to make a PR to setuptools for adding wheels support to `easy_install`, which is the base of `zc.recipe.egg`'s ability to install packages.

It's important to remember that [many](https://pypi.python.org/pypi/anybox.recipe.odoo/) [buildout](https://pypi.python.org/pypi/djangorecipe) [recipes](https://pypi.python.org/pypi/plone.recipe.zeoserver) are based on `zc.recipe.egg` (either by inheritance or by delegation), so even though it would not be too difficult to create a new buildout recipe that can handle wheels, either by shelling out to `pip` or by using something like `distlib`, until `zc.recipe.egg` can handle wheels by itself, all these other recipes will still miss out on wheels.

And with the manylinux project, the lack of wheel support is starting to hurt too close to home.

**Comment by pombredanne** [comment]
@leorochael You should ask @jaraco and @dholth if there is any plan to do wheels install in setuptools. 

I think this would be a tad too much overlap and rework between pip and setuptools IMHO... 

As for me, I shall say that after I contributed the code to convert buildout eggs to wheels (which has been pulled since wheel 0.24 and up) I stopped using buildout (sadly) in part because of wheels and pip giving me a more consistent and faster installation experience (in part likely because of pip caching and such, and in part because some hard to build wheels are provided pre-built on Pypi) and in part because I could not get the install-from-offline-vendored-eggs patches I had contributed to buildout stay there after they were pulled in. In contrast installing from vendored wheels has been fast and consistent for me. And I created my poor mans buildout replacement with this script https://github.com/nexB/scancode-toolkit/blob/develop/etc/configure.py that I use in many projects.

**Comment by dholth** [comment]
Personally I am not interested in adding wheel to setuptools. I am more interested in subtracting setuptools and distutils and even bdist_wheel itself from my own packages using enscons. But that is the packaging side, not the installing side.

It should be easy to add wheel install support to setuptools or easy_install or whatever buildout may need. pkg_resources was the first tool to support dist-info directories, and once they are unpacked they look exactly like eggs in your pkg_resources.working_set. There's code from the 'wheel install' command that could be called and is fairly setuptools-onic. Use wheel.install.WheelFile() instances which are sortable by how desirable the particular wheel is, this works with only the filename. Grab WheelFile.install_paths(), change them all to point to the single directory in which it should be installed egg-style, pass to WheelFile.install(overrides={...}), then generate console scripts as usual.

The only part that might need a little love is wheel.pep425tags which generates a slightly different list of supported tags than pip.pep425tags, mainly pip supports more OSX tags. They should probably just be kept identical.

**Comment by leorochael** [comment]
@dholth thanks for the info, I'll copy it to the [`setuptools` issue](https://github.com/pypa/setuptools/issues/78), since it's more relevant there.

I just pinged people here in case anyone else has some time they would be willing to spend on `setuptools` to make it happen.

**Comment by jimfulton** [comment]
I made progress on this today.

I realized that with wheels, buildout could potentially install things like numpy whose builds depended on being installed into a real Python directory tree.

Looking at the comments here and in the setuptools thread, I was able to add a small buildout hook and create an extension that adds wheel support.  I was able to use it to install numpy with buildout and no pip.

I need to write tests. Stay tuned for a PR and an extension, hopefully, tomorrow.

**Comment by leorochael** [comment]
> Looking at the comments here and in the setuptools thread, I was able to add a small buildout hook and create an extension that adds wheel support. I was able to use it to install numpy with buildout and no pip.

Is this support added directly to setuptools? Or are you doing an end run on setuptools to install wheels directly?

**Comment by jimfulton** [comment]
On Mon, Feb 13, 2017 at 7:03 AM, Leonardo Rochael Almeida <
notifications@github.com> wrote:

> Looking at the comments here and in the setuptools thread, I was able to
> add a small buildout hook and create an extension that adds wheel support.
> I was able to use it to install numpy with buildout and no pip.
>
> Is this support added directly to setuptools? Or are you doing an end run
> on setuptools to install wheels directly?
>

End run for now. There are two parts:

- A small monkey patch (done by a buildout extension) to make setuptools
treat wheels as source distros and parse their file names correctly,

- And a hook in buildout to be provided by an extension, to convert wheels
to eggs.

This is all appropriately hackish right now.  :) The extension depends on
wheel and a the "humpty" package to do the conversion, which brings in lots
of other packages.  I had to work around a bug in distlib, which humpty
uses, that misinterpreted wheel names on Mac.

As an initial step, I want to get something out there that works.

Then, I can see replacing humpty with something much leaner.

Then maybe pushing back into setuptools, however, there's an inherent egg
limitation that may make this awkward for setuptools.  Egg names don't
include abi data included in wheels and this information is lost in the
conversion. I'm not sure is this is a problem is eggs are only used as an
installation format.  IN any case, this isn't a problem for buildout thanks
to the recently-added option to use separate eggs subdirectories for
different abi tags.

Jim

-- 
Jim Fulton
http://jimfulton.info

**Comment by leorochael** [comment]
> End run for now. There are two parts:
> 
> - A small monkey patch (done by a buildout extension) to make setuptools
> treat wheels as source distros and parse their file names correctly,
> 
> - And a hook in buildout to be provided by an extension, to convert wheels
> to eggs.
> 
> This is all appropriately hackish right now.  :) The extension depends on
> wheel and a the "humpty" package to do the conversion, which brings in lots
> of other packages.  I had to work around a bug in distlib, which humpty
> uses, that misinterpreted wheel names on Mac.

This sounds a lot more work than getting setuptools to install wheels directly. From the point of view of setuptools, wheels look a lot like bdist_wininst installers: zip packages that can have installable files starting from the root, but also starting from some other sub-directories depending on whether the wheel has a platform specific files inside or not.

> Then maybe pushing back into setuptools, however, there's an inherent egg limitation that may make this awkward for setuptools.  Egg names don't include abi data included in wheels and this information is lost in the conversion.

Converting from `wheel` to `egg` is indeed "lossy". In practice this is not a problem if we're not going to try to distribute the generated eggs.

> I'm not sure is this is a problem is eggs are only used as an installation format.

Specifically, for buildout, unzipped eggs are a hybrid of installation and caching format. A way of reusing the same set of files in multiple places without taking additional space.

This is one feature I'm not seeing anywhere else, and it makes me really sad that no one else is looking into it, and we run the risk of throwing it out with the bathwater in a future where `buildout` only automates `pip` (of course, this not a problem for your current approach.

> IN any case, this isn't a problem for buildout thanks to the recently-added option to use separate eggs subdirectories for different abi tags.

Exactly.

**Comment by jimfulton** [comment]
On Mon, Feb 13, 2017 at 9:32 AM, Leonardo Rochael Almeida <
notifications@github.com> wrote:

> End run for now. There are two parts:
>
>    -
>
>    A small monkey patch (done by a buildout extension) to make setuptools
>    treat wheels as source distros and parse their file names correctly,
>    -
>
>    And a hook in buildout to be provided by an extension, to convert
>    wheels
>    to eggs.
>
> This is all appropriately hackish right now. :) The extension depends on
> wheel and a the "humpty" package to do the conversion, which brings in lots
> of other packages. I had to work around a bug in distlib, which humpty
> uses, that misinterpreted wheel names on Mac.
>
> This sounds a lot more work than getting setuptools to install wheels
> directly. From the point of view of setuptools, wheels look a lot like
> bdist_wininst installers: zip packages that can have installable files
> starting from the root, but also starting from some other sub-directories
> depending on whether the wheel has a platform specific files inside or not.
>
Go for it. :)

There's sooo much weird caching and indirection in the setuptools code
base, it's a nightmare figuring what the heck is going on. I feel like I'm
in Maine and wherever I am, I "can't get there from here". :)

Also, by making this a buildout extension, I was able to leverage other
code without worrying about adding dependencies.

So far, to make this work, I've added 6 lines to buildout (including a
blank line :)) and the extension has 66 source lines (not counting
administrative files, like setup.py, etc.).  This was possible because I'm
depending on and reusing the wheel and humpty packages.  Of course, I'm
monkey patching setuptools, which has risks, and my source-line count will
go up a lot when I add tests.

And of course, I don't have to worry about making a setuptools PR and
getting it accepted.

I'm certainly in favor of someone getting this, or at least some of it back
into setuptools, at least to the point where we can avoid the monkey
patches.

Jim

-- 
Jim Fulton
http://jimfulton.info

**Comment by jimfulton** [comment]
https://github.com/buildout/buildout/pull/332

**Comment by jimfulton** [comment]
Initial extension implementation: https://github.com/buildout/wheel

I'll be adding tests before an initial release.

**Comment by jimfulton** [comment]
buildout.wheel is out:

   https://pypi.python.org/pypi/buildout.wheel

**Comment by jimfulton** [comment]
If someone wants to help, it would be great to get rid of humpty (and especially distlib) use by the extension.  

Distlib seems to be an opportunity to fail. (It fails on the python-dateutil wheel, for example, which prevented me from installing pandas in absence of existing python-dateutil eggs.)

This would be a step toward making a setuptools PR more likely.

**Comment by leorochael** [comment]
Dropping an idea of how to get rid of humpty here so that I can find it later:

 * Use `wheel` to install it to a specific directory as if it was `site-packages`, name this directory `equivalent-egg-name-of-wheel-<version>.egg`
 * Walk up the namespace packages in that directory declared in the wheel to make sure they contain proper `__init__` with namespace declarations (to fix the issue in [this comment](https://github.com/buildout/buildout/issues/319#issuecomment-271017185))

**Comment by aclark4life** [comment]
TIL about [Humpty](https://github.com/dairiki/humpty/) 😄

**Comment by jimfulton** [comment]
On Thu, Feb 16, 2017 at 4:00 PM, Leonardo Rochael Almeida <
notifications@github.com> wrote:

> Dropping an idea of how to get rid of humpty here so that I can find it
> later:
>
>    - Use wheel to install it to a specific directory as if it was
>    site-packages, name this directory equivalent-egg-name-of-wheel-<
>    version>.egg
>
> That probably won't handle scripts correctly.

>
>    - Walk up the namespace packages in that directory declared in the
>    wheel to make sure they contain proper __init__ with namespace
>    declarations (to fix the issue in this comment
>    <https://github.com/buildout/buildout/issues/319#issuecomment-271017185>
>    )
>
> Oh gawd.  I'm going to try to ignore that. :)

Note that not all namespace packages packages should include __init__
scripts.
Including one could have the effect of breaking the namespace:

  https://www.python.org/dev/peps/pep-0420/#specification

You also have to create egg-info directories.

IMO, it would be probably be simpler to just unpack the wheel and add just
enough egg-info for buildout to work, being especially careful to handle
requirements and non-entry-point scripts correctly.

**Comment by leorochael** [comment]
> >   - Use wheel to install it to a specific directory as if it was site-packages, name this directory equivalent-egg-name-of-wheel-<version>.egg
> 
>  That probably won't handle scripts correctly.

[Entry-point `console_scripts`](http://python-packaging.readthedocs.io/en/latest/command-line-scripts.html#the-console-scripts-entry-point) will be handled correctly if we get the `egg/dist-info` right (see below).

[`scripts=` entries](http://python-packaging.readthedocs.io/en/latest/command-line-scripts.html#the-scripts-keyword-argument) is what I'm not so sure about. Questions at the end.

> Note that not all namespace packages packages should include __init__ scripts.
> Including one could have the effect of breaking the namespace:
> 
>  https://www.python.org/dev/peps/pep-0420/#specification

According to that spec, including an `__init__.py` will stop a directory from being implicitly treated as an namespace package. But if that `__init__.py` contains itself a namespace declaration, then the namespace declaration will make that package name into an "explicit" namespace package.

According to the PEP you linked above:

> Namespace packages cannot contain an `__init__.py`. As a consequence, `pkgutil.extend_path` and `pkg_resources.declare_namespace` become _obsolete_ for purposes of namespace package creation.

But it doesn't say they were going to be removed or forbidden. In fact `pkgutil.extend_path()` is still documented, and not marked as obsolete, in the [Python 3 stdlib docs](https://docs.python.org/3/library/pkgutil.html#pkgutil.extend_path).

Which is how namespace packages that straddle Python 2 and 3 handle namespaces, at least in _develop_ mode.

So we would have two options:

 * Add missing `__init__.py` namespace package declarations for Python 2 only.
 * Add missing `__init__.py` namespace package declarations always until Python 3 drops support for `pkgutil.extend_path()`.

> You also have to create egg-info directories.

According to some tests I did after a tip from someone I forgot now, while discussing about adding wheel support to `setuptools`, we don't actually need to create `egg-info` directories.

`pkg_resources` can parse directories named `dist_package_name-version.dist-info` that happen to be in `sys.path` just as well as it can parse `.egg-info` directories. 

`dist-info` directories already present inside the wheel, with all necessary info (incl. setuptools entry points). You just need to drop it in `sys.path` somewhere. For example, a directory that just so happens to be called `dist_package_name-version.egg` and is, for some coincidence, present in the `sys.path` by pure luck ;-)

> IMO, it would be probably be simpler to just unpack the wheel

So, basically, just unpack the wheel, moving the platform/architecture specific subdirectories of the wheel to the root of the unpacked tree, and make sure this unpacked tree ends up in `sys.path` somehow, which will magically place the `.dist-info` directory where we need it.

> and add just enough egg-info for buildout to work

Which is none, considering the `dist-info` thing I mentioned above.

> being especially careful to handle requirements and non-entry-point scripts correctly.

This is the part where I'm fuzzy on the details. How does buildout deal with it non-entry-pont scripts today? If the answer is "not at all", do we need to care about them when they come from wheels?

**Comment by jimfulton** [comment]
On Fri, Feb 17, 2017 at 11:42 AM, Leonardo Rochael Almeida <
notifications@github.com
<https://mail.google.com/mail/?view=cm&fs=1&tf=1&to=notifications@github.com>
> wrote:

>
>    - Use wheel to install it to a specific directory as if it was
>    site-packages, name this directory equivalent-egg-name-of-wheel-.egg
>
> That probably won't handle scripts correctly.
>
> Entry-point console_scripts
> <http://python-packaging.readthedocs.io/en/latest/command-line-scripts.html#the-console-scripts-entry-point>
> will be handled correctly if we get the egg/dist-info right (see below).
>

Yup. I was mainly referring to non-entry-point scripts.


> scripts= entries
> <http://python-packaging.readthedocs.io/en/latest/command-line-scripts.html#the-scripts-keyword-argument>
> is what I'm not so sure about. Questions at the end.
>
> Note that not all namespace packages packages should include *init*
> scripts.
> Including one could have the effect of breaking the namespace:
>
> https://www.python.org/dev/peps/pep-0420/#specification
>
> According to that spec, including an __init__.py will stop a directory
> from being implicitly treated as an namespace package. But if that
> __init__.py contains itself a namespace declaration, then the namespace
> declaration will make that package name into an "explicit" namespace
> package.
>
Right. But if other members of the namespace were installed without
__init__.py files (perhaps because non-wheel distros) then they will
effectively be ejected from the namespace. That's the breakage I was
referring to.



> According to the PEP you linked above:
>
> Namespace packages cannot contain an __init__.py. As a consequence,
> pkgutil.extend_path and pkg_resources.declare_namespace become *obsolete*
> for purposes of namespace package creation.
>
> But it doesn't say they were going to be removed or forbidden. In fact
> pkgutil.extend_path() is still documented, and not marked as obsolete, in
> the Python 3 stdlib docs
> <https://docs.python.org/3/library/pkgutil.html#pkgutil.extend_path>.
>
Right. I expect old style and new style to be allowed.  My point is that
adding an __init__ is likely to break things.
(It feels brittle that old style and new-style can't be mixed, AFAICT.)



> Which is how namespace packages that straddle Python 2 and 3 handle
> namespaces, at least in *develop* mode.
>
> So we would have two options:
>
>    - Add missing __init__.py namespace package declarations for Python 2
>    only.
>    - Add missing __init__.py namespace package declarations always until
>    Python 3 drops support for pkgutil.extend_path().
>
> Why add them at all? If they're needed, why wouldn't they have been
included in the wheel?



>
>
> You also have to create egg-info directories.
>
> According to some tests I did after a tip from someone I forgot now, while
> discussing about adding wheel support to setuptools, we don't actually
> need to create egg-info directories.
>
> pkg_resources can parse directories named dist_package_name-version.
> dist-info that happen to be in sys.path just as well as it can parse
> .egg-info directories.
>
Ah, cool. Does that handle scripts as well? (You don't have to answer now,
I'm sure we'll find out. :)



> dist-info directories already present inside the wheel, with all
> necessary info (incl. setuptools entry points). You just need to drop it in
> sys.path somewhere. For example, a directory that just so happens to be
> called dist_package_name-version.egg and is, for some coincidence,
> present in the sys.path by pure luck ;-)
>

Cool. I hope our luck holds. :)

One thing that helps our luck is that we don't need a full-fidelity
conversion. We just need the generated eggs to work as an installation
format.


> IMO, it would be probably be simpler to just unpack the wheel
>
>
Notice my weasel words. :)

> So, basically, just unpack the wheel, moving the platform/architecture
> specific subdirectories of the wheel to the root of the unpacked tree, and
> make sure this unpacked tree ends up in sys.path somehow, which will
> magically place the .dist-info directory where we need it.
>
The result should look like an egg, so have a valid file name, etc.

> and add just enough egg-info for buildout to work
>
> Which is none, considering the dist-info thing I mentioned above.
>

Hopefully.


> being especially careful to handle requirements and non-entry-point
> scripts correctly.
>
> This is the part where I'm fuzzy on the details. How does buildout deal
> with it non-entry-pont scripts today? If the answer is "not at all", do we
> need to care about them when they come from wheels?
>
It does handle them. Grep for _distutils_script in the sources.  In eggs
they're in an egg-info sub-directory.

My advice:

- Initially, approach this in discovery/hack mode.  Initially, try whatever
you guess us easiest, whether that's by instaling with wheel, unpacking the
wheel.  See what seems to work. It doesn't matter if you guess wrong.

- Use humpty output as a reference for things you're unsure of.

- I've been using numpy as a test case because it was impossible to install
it with buildout before. It happens to include a non-entry-point script.

**Comment by leorochael** [comment]
> - I've been using numpy as a test case because it was impossible to install it with buildout before. It happens to include a non-entry-point script.

Funny, I regularly install numpy with buildout. Takes a while to compile, and requires quite a few O.S. dependencies (*cough*, *fortran*, *cough*), but a global buildout egg cache usually takes care of it.

It's [`pandas` that has been giving me trouble recently](https://github.com/buildout/buildout/issues/313).

**Comment by leorochael** [comment]
> Why add them at all? If they're needed, why wouldn't they have been included in the wheel?

Take for example `zest.releaser`:

Install it with `pip` (to use the wheel instead of the source) into a virgin virtualenv and you'll have a `site-packages/zest` that is empty except for the `releaser` directory.

On Python 3, that just works, due to [PEP-420](https://www.python.org/dev/peps/pep-0420).

On Python 2, it shouldn't work. It only does because of a little turd dropped by `pip` called `site-packages/zest.releaser-6.9-py2.7-nspkg.pth`:

    import sys, types, os;p = os.path.join(sys._getframe(1).f_locals['sitedir'], *('zest',));ie = os.path.exists(os.path.join(p,'__init__.py'));m = not ie and sys.modules.setdefault('zest', types.ModuleType('zest'));mp = (m or []) and m.__dict__.setdefault('__path__',[]);(p not in mp) and mp.append(p)

The little turd above will not be taken into account if it's just in a folder on `sys.path`. To be taken into account, it would have to be in a `site` folder. I.e. a folder on which we called [`site.addsitedir()`](http://stackoverflow.com/a/41753634/621201)

If some people scream at buildout because it adds a ton of stuff to `sys.path`, imagine the reaction of adding all buildout eggs of a part into `sitedir`...

**Comment by leorochael** [comment]
> On Python 2, it shouldn't work. It only does because of a little turd dropped by pip called `site-packages/zest.releaser-6.9-py2.7-nspkg.pth`:

My mistake, it's not pip that does this. The file is in `zest.releaser` itself, and `humpty` gets rid of it and populates the namespace folders with namespace-declaring `__init__.py` files.

**Comment by jimfulton** [comment]
On Fri, Feb 17, 2017 at 2:00 PM, Leonardo Rochael Almeida <
notifications@github.com> wrote:

> Why add them at all? If they're needed, why wouldn't they have been
> included in the wheel?
>
> Take for example zest.releaser:
>
> Install it with pip (to use the wheel instead of the source) into a
> virgin virtualenv and you'll have a site-packages/zest that is empty
> except for the releaser directory.
>
That's weird. The source has an __init__.py:

*https://github.com/zestsoftware/zest.releaser/tree/master/zest
<https://github.com/zestsoftware/zest.releaser/tree/master/zest>*

The setup.py is declaring zest to be a namespace package. I wonder is pip
is using that meta-data.   This feels sooooo brittle.

Oh gawd....

I hope pip does the same thing for something that's just a source release.

I wonder what it does with "editable" installs.

This feels like a minefield.

Jim

-- 
Jim Fulton
http://jimfulton.info

**Comment by jimfulton** [comment]
On Fri, Feb 17, 2017 at 2:06 PM, Leonardo Rochael Almeida <
notifications@github.com> wrote:

> On Python 2, it shouldn't work. It only does because of a little turd
> dropped by pip called site-packages/zest.releaser-6.9-py2.7-nspkg.pth:
>
> My mistake, it's not pip that does this. The file is in zest.releaser
> itself, and humpty gets rid of it and populates the namespace folders
> with namespace-declaring __init__.py files.
>
sooooooo much magic.

**Comment by mgedmin** [comment]
On the topic of mixing PEP-420 namespace packages with old-style magic namespace packages, there's an upstream issue filed (https://github.com/pypa/setuptools/issues/912).  It's still not clear to me if that's allowed or not, and why it sometimes works, and sometimes doesn't.

**Comment by delijati** [comment]
This can be closed as of https://github.com/buildout/buildout/issues/319#issuecomment-384923839

**Comment by mauritsvanrees** [comment]
Wheels are supported since long.
